File: objc/NSOperationObjCImageLoader/NSOperationObjCImageLoader/PendingOperations.m
//
//  PendingOperations.m
//  NSOperationObjCImageLoader
//
//  Created by zhen gong on 8/27/17.
//  Copyright © 2017 zhen gong. All rights reserved.
//

#import "PendingOperations.h"

@implementation PendingOperations

- (instancetype)init {
    if(self = [super init]) {
        _map = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)updateMapWithKey:(NSString *)key {
    NSNumber* count = [_map valueForKey:key];
    if(count == nil) {
        [_map setValue:@1 forKey:key];
    } else {
        int countInt = [count intValue];
        [_map setValue:@(countInt + 1) forKey:key];
    }
}

- (NSOperationQueue *)downloadQueue {
    if(_downloadQueue == nil) {
        _downloadQueue = [[NSOperationQueue alloc] init];
        _downloadQueue.name = @"Download queue";
        _downloadQueue.maxConcurrentOperationCount = 1;
    }
    return _downloadQueue;
}

- (NSMutableDictionary *)downloadsInProgress {
    if(_downloadsInProgress == nil) {
        _downloadsInProgress = [NSMutableDictionary new];
    }
    return _downloadsInProgress;
}

@end

File: swift/NSOperationSwift/NSOperationSwift/NSOperation/PendingOperations.swift
//
//  PendingOperations.swift
//  NSOperationSwift
//
//  Created by zhen gong on 6/30/17.
//  Copyright © 2017 zhen gong. All rights reserved.
//

import Foundation

class PendingOperations{
    
    var map:[String: Int] = [String: Int]()
    
    func updateMap(key:String) {
        if let value = map[key] {
            map.updateValue(value + 1, forKey: key)
        } else {
            map.updateValue(1, forKey: key)
        }
    }
    
    
//    var map:[String: Int]? {
//        willSet {
//        
//        }
//        didSet {
//        
//        }
//    }
//    
    lazy var downloadsInProgress = [IndexPath:Operation]()
    lazy var downloadQueue: OperationQueue = {
        var queue = OperationQueue()
        queue.name = "Download queue"
        queue.maxConcurrentOperationCount = 1
        return queue
    }()
    
    lazy var filtrationsInProgress = [IndexPath:Operation]()
    lazy var filtrationQueue: OperationQueue = {
        var queue = OperationQueue()
        queue.name = "Image Filtration queue"
        queue.maxConcurrentOperationCount = 1
        return queue
    }()
    
}

